### PR TITLE
Notations for numerals are treated natively and do not modify the constr parser

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

Depends on #14070 (merged) and #14071 (merged). Edit: depends on #14562 (merged).

This PR was originally in #13353 but it happens to be a significant change enough to grow as itself and be in its own PR.

For instance:
```coq
Notation "0" := tt (at level 3).
Check 0%nat.
```
should not return `tt`.

So we simply forbid notations to numerals to modify the parser. It is currently implemented as a warning telling that the modifiers are discarded but it could also be an error. The warning seems better since it would let developments that rely on it to be changed at their own pace.

Note that numerals can still be used in notations but no new rule is added for them in the parser.

Further additions (25/6/21):

```
Notation "0 +" := None (format "0 +").
Check 0. (* was blocked by the "0 +" notation before *)
```

```
Notation "0+" := None.
Check 0+. (* was working before for printing but not for parsing *)
```

- [X] Added / updated test-suite
- [ ] Entry added in the changelog (is it worth or too technical?)

### CODE REVIEW

**Comment by proux01** [comment]
@herbelin does that mean that literals such as "0" or "1" shouldn't be used as non-number notations? That lind of use seems not that uncommon, for instance in mathcomp. Sure, a number notation can always be used but that's slightly more heavy to setup. Printing of number notations / interaction with printing of regular notations is certainly still buggy but I wonder whether this is the best fix?

**Comment by herbelin** [comment]
> @herbelin does that mean that literals such as "0" or "1" shouldn't be used as non-number notations? That kind of use seems not that uncommon, for instance in mathcomp. Sure, a number notation can always be used but that's slightly more heavy to setup. Printing of number notations / interaction with printing of regular notations is certainly still buggy but I wonder whether this is the best fix?

Sorry for so late answer. `0`or `1` can still be used as notations for arbitrary terms. The only change is that this does not declare a new parsing rule. It is now explicitly a pure interpretation thing. In particular,
```
Notation "0" := tt (at level 3).
```
is still accepted, but it does not hide anymore the standard parsing rule for `0` as a numeral.

For instance, I did not check but I suspect that redeclarations of numeral rules such as done in [fourcolour](https://github.com/herbelin/fourcolor/blob/a72978a10c1eace75c2d7bb37b87f6cf7a7a4d69/theories/part.v#L170-L173) would not be necessary anymore.

**Comment by silene** [comment]
What is the status with respect to #14071? Is it still needed or is #14429 sufficient as a dependency?

**Comment by herbelin** [comment]
It still depend on #14071 (but not on #14429).

**Comment by proux01** [comment]
Given that #14071 got postponed to 8.15, I guess this should be to. I'll try to take soon a closer look though.

File: coqpp/coqpp_parse.mly
Comment (proux01) [inline coqpp/coqpp_parse.mly:396]: This commit should probably be removed from this PR considering the discussion in https://github.com/coq/coq/pull/14071#discussion_r642929069

**Comment by proux01** [comment]
@herbelin could you please rebase now that dependency PRs are merged, in order to ease review?

**Comment by herbelin** [comment]
> @herbelin could you please rebase now that dependency PRs are merged, in order to ease review?

I rebased but also made the PR evolved. First, it has a new dependency #14562 (present here in the 2 first commits).

Second, it addresses 2 new cases of bad interactions in the presence of notations with numbers.

File: coqpp/coqpp_parse.mly
Comment (herbelin) [reply coqpp/coqpp_parse.mly]: Done.

File: doc/tools/docgram/orderedGrammar
Comment (jfehrle) [inline doc/tools/docgram/orderedGrammar:93]:
```suggestion
| number
| string
```

Better to keep the existing order here with number first since the doc discusses number and string in that order.  Then please run `make doc_gram_rsts` to restore that order in the rst file.  Doc_grammar is unable to preserve the existing order for certain changes to the grammar (a defect in the mind-reading part of the code).

**Review by proux01 — APPROVED** [review]
Sorry for the very late review.
This looks good to me, I'll merge at the end of the week if no further comments.

**Comment by proux01** [comment]
@coqbot merge now

**Comment by herbelin** [comment]
@proux01: you probably have to join the @coq/extensible-syntax-maintainers (cc @Zimmi48). Maybe could we also create a group of maintainers on numbers in general (number notations, int, float, ...).

**Comment by ppedrot** [comment]
Let me merge in the meantime, there is no point in leaving this open.

**Comment by proux01** [comment]
Thanks!

**Comment by Zimmi48** [comment]
Sorry for not responding earlier.

@proux01 I could add you right now to the `@coq/extensible-syntax-maintainers` team or wait until the `@coq/number-maintainers` team is created. To do the latter, please open a PR updating https://github.com/coq/coq/blob/master/.github/CODEOWNERS with the list of files the team should own (note that order matters in this file). @JasonGross might also be interested in being a member of this new team.

Also, let me know if there are any other maintainer teams (see the list at https://github.com/orgs/coq/teams/pushers/teams) that you'd like to join (e.g., `@coq/vm-native-maintainers`).

**Pull request merged**
